### PR TITLE
AFGL 1986 standard radiation atmospheres

### PR DESCRIPTION
**Pull request by nollety**
# Description

Resolves eradiate/eradiate-issues#44, eradiate/eradiate-issues#52, eradiate/eradiate-issues#53 and eradiate/eradiate-issues#54.

Changes:
- I've added a `AFGL1986RadProfile` class to `eradiate/radprops/rad_profile.py`.
- The data submodule was updated to includes the `profiles/` and the `tests/spectra/absorption`.
- I've updated `eradiate/data` to include a profiles getter.
- I've added a `thermoprops/afgl_1986.py` module to make the corresponding AFGL 1986 atmospheric profiles.
- I've updated `thermoprops/util.py` to support vertical and concentration rescaling.
- I've updated all `RadProfile` classes with a `levels` attribute (array) to support vertical rescaling of the underlying thermophysical properties profile.
- I've added `eradiate/data/chemistry` to include a chemical data sets getter.

# Checklist

- [x] The code follows the relevant coding guidelines
- [x] The code generates no new warnings
- [x] The code is appropriately documented
- [x] The code is tested to prove its function
- [x] The feature branch is rebased on the current state of the `main` branch
- [x] I give permission that the Eradiate project may redistribute my contributions under the terms of its license

### CODE REVIEW

**Review by leroyvn — COMMENTED** [review]
Hi @nollety, it's not clear to me just yet: does this mean that the feature can actually not be used yet? I assume so since I don't see tests. If so, I'd rather delay the review and merge to when it will be ready.

**Comment by nollety** [comment]
@leroyvn, that's right, the feature cannot be used yet. Alright, let's wait!

**Comment by nollety** [comment]
@leroyvn I'm in the process of creating the "dev"-absorption data sets:

- [ ] CH4
- [ ] CO
- [ ] CO2
- [ ] H2O
- [ ] N2O
- [ ] O2
- [ ] O3

**Comment by nollety** [comment]
In the meantime I'll merge the content of #41 in this pull request

**Comment by nollety** [comment]
The dev absorption data sets are ready, I'm running some tests on them and then I upload them to the eradiate ftp server.

**Comment by nollety** [comment]
Preliminary tests passed: I started uploading the data sets to the ftp server.

**Comment by nollety** [comment]
Upload completed: total size = 23 GB.

**Comment by nollety** [comment]
I discovered an error in `eradiate.radprops.absorption.compute_sigma_a`; the multi-dimensional interpolation is not performed correctly: I'm fixing that.

File: eradiate/data/__init__.py [hunk 1]
```diff
@@ -40,12 +42,14 @@
...
 
 from .absorption_spectra import _AbsorptionGetter
+from .profiles import _ProfilesGetter
 from .solar_irradiance_spectra import _SolarIrradianceGetter
 from .spectral_response_function import _SpectralResponseFunctionGetter
 from .. import path_resolver as _presolver
 
 _getters = {
     "absorption_spectrum": _AbsorptionGetter,
+    "profiles": _ProfilesGetter,
```
Comment (leroyvn) [inline eradiate/data/__init__.py:52]: Same here: please use `atmospheric_profiles`.

**Review by leroyvn — COMMENTED** [review]
Hi @nollety, thanks for this (big) PR. I made a few comments about style, which I could summarise with the following:

* please apply Black and isort to your files;
* please start your docstrings with a blank line.

The interfaces and module organisation you put together comply with the current state of `main` but I'll change some things after merging as part of kernel dict context implementation. Other than that, I don't have much to say in general (see specific points).

File: eradiate/data/tests/test_data.py [hunk 2]
```diff
@@ -33,4 +33,3 @@ def test_required_data_sets():
     # Check that data sets required for testing are found
     assert data.find("solar_irradiance_spectrum")["thuillier_2003"]
     assert data.find("solar_irradiance_spectrum")["blackbody_sun"]
-    assert data.find("absorption_spectrum")["test"]
```
[the remark this answers is not in the document]
Comment (nollety) [reply eradiate/data/tests/test_data.py:36]: Yes, I figured we can simply create a test data set object with `xr.Dataset()` and that's all we need as far as `test_absorption.py` is concerned.

File: eradiate/radprops/rad_profile.py [hunk 3]
```diff
@@ -468,101 +483,108 @@ class US76ApproxRadProfile(RadProfile):
...
-        attr.ib(
-            default=50,
-            converter=int,
-            validator=is_positive
-        ),
-        doc="Number of atmospheric layers. Default: 50",
-        type="int",
-    )
-
-    height = documented(
```
[the remark this answers is not in the document]
Comment (nollety) [reply eradiate/radprops/rad_profile.py:481]: Yes, exactly. This is the interface that Yves requested. Computing `height` and `n_layers` from `levels` is easy of course:
```python
height = levels.values.max()
n_layers = len(levels)
```
Vertical rescaling is supported naturally with this interface.
Also, this kinda solves the issue that `n_layers=32` is not understood by typical RTM users to mean 32 equal-width atmospheric layers, but instead some (reference unknown to me just yet) standard altitude mesh with 32 non-equal-width layers. Now, we can do something like that:
```python
REFERENCE_4_LAYERS = np.array(0, 10, 20, 50, 120)
profile = AFGL1986RadProfile(levels=REFERENCE_4_LAYERS)
```
as soon as that reference is identified.

File: eradiate/radprops/rad_profile.py [hunk 4]
```diff
@@ -601,15 +623,252 @@ def sigma_t(self):
...
+    "midlatitude_winter",
+    "subarctic_summer",
+    "subarctic_winter",
+    "us_standard"]
+
+
+@RadProfileFactory.register("afgl1986")
+@parse_docs
+@attr.s
+class AFGL1986RadProfile(RadProfile):
```
[the remark this answers is not in the document]
Comment (nollety) [reply eradiate/radprops/rad_profile.py:636]: Okay!

File: eradiate/data/__init__.py (hunk 1 above)
Comment (nollety) [reply eradiate/data/__init__.py:52]: After our discussion, I went for `thermoprops_profiles`.

File: docs/rst/reference/data.rst [hunk 5]
```diff
@@ -23,6 +23,7 @@ Data sets
    :recursive:
    :toctree: generated/
 
-   solar_irradiance_spectra
    absorption_spectra
+   profiles
```
[the remark this answers is not in the document]
Comment (nollety) [reply docs/rst/reference/data.rst:27]: After our discussion, I went for `thermoprops_profiles`.

**Comment by nollety** [comment]
@leroyvn I'm done with the generation of the (extremely light) test absorption data sets ; I've added them to eradiate-data. Now, I need to adapt a bit the interface of US76ApproxRadProfile and AFGL1986RadProfile classes so that I can test them with the test absorption data sets. Hopefully, I'll be done with that tomorrow!

**Comment by nollety** [comment]
@leroyvn This is ready for another review :)

**Comment by nollety** [comment]
Thanks to the reduced absorption test data sets, tests duration has decreased significantly:
`radprops` tests: < 0.6 s
`scenes/atmosphere tests`: < 0.4 s

**Comment by nollety** [comment]
Note: I would like to add more tests, especially of the rescaling features. I want to check thoroughly that molecular species concentrations are correctly computed, that the atmosphere thermophysical properties are correctly rescaled with these "rescaling factors" and that the resulting absorption coefficient is correctly computed as the weighted sum of the individual molecular species absorption coefficient. However, I'll wait for your feedback before adding more tests, just in case a lot of changes are required.

File: eradiate/thermoprops/util.py [hunk 6]
```diff
@@ -1,27 +1,339 @@
-"""Utility functions to manipulate atmospheric profiles. """
+"""Utility functions to manipulate atmosphere thermophysical properties data
```
Comment (leroyvn) [inline eradiate/thermoprops/util.py:1]:
```suggestion
"""
Utility functions to manipulate atmosphere thermophysical properties data 
```

File: eradiate/thermoprops/util.py [hunk 7]
```diff
@@ -1,27 +1,339 @@
-"""Utility functions to manipulate atmospheric profiles. """
+"""Utility functions to manipulate atmosphere thermophysical properties data 
+sets. """
```
Comment (leroyvn) [inline eradiate/thermoprops/util.py:2]:
```suggestion
sets.
"""
```

File: eradiate/thermoprops/util.py [hunk 8]
```diff
@@ -1,27 +1,339 @@
...
-    var_specs={
-        "p": VarSpec(standard_name="air_pressure", units="Pa", long_name="air pressure"),
-        "t": VarSpec(standard_name="air_temperature", units="K", long_name="air temperature"),
-        "n": VarSpec(standard_name="air_number_density", units="m^-3", long_name="air_number density"),
-        "mr": VarSpec(standard_name="mixing_ratio", units="", long_name="mixing ratio"),
-    },
-    coord_specs="atmospheric_profile"
-)
+def compute_column_number_density(ds, species):
+    """Computes the column number density of a given species in an atmospheric
```
Comment (leroyvn) [inline eradiate/thermoprops/util.py:22]:
```suggestion
    """
    Computes the column number density of a given species in an atmospheric
```

File: eradiate/thermoprops/util.py [hunk 9]
```diff
@@ -1,27 +1,339 @@
...
+            raise ValueError(
+                f"Invalid dimension for {species} concentration:" f" {amount.units}."
+            )
+        factors[species] = factor.magnitude
+
+    return factors
+
+
+def human_readable(items):
+    """Transforms a list into readable human text.
```
Comment (leroyvn) [inline eradiate/thermoprops/util.py:220]:
```suggestion
    """
    Transforms a list into readable human text.
```

File: eradiate/thermoprops/util.py [hunk 10]
```diff
@@ -1,27 +1,339 @@
...
+    x = f"{items[0]}"
+    for s in items[1:-1]:
+        x += f", {s}"
+    if len(items) > 1:
+        x += f" and {items[-1]}"
+    return x
+
+
+def rescale_concentration(ds, factors, inplace=False):
+    """Rescale mixing ratios in an atmosphere thermophysical properties data
```
Comment (leroyvn) [inline eradiate/thermoprops/util.py:239]:
```suggestion
    """
    Rescale mixing ratios in an atmosphere thermophysical properties data
```

File: eradiate/thermoprops/util.py [hunk 11]
```diff
@@ -1,27 +1,339 @@
...
+        f"Concentration-rescaled for {human_readable(species)} - "
+        f"eradiate.thermoprops.util.rescale_concentration"
+    )
+
+    return ds
+
+
+@ureg.wraps(ret=None, args=(None, "m", None, None), strict=False)
+def interpolate(ds, z_level, method="linear", conserve_columns=False):
+    """Interpolates an atmosphere thermophysical properties data set onto a
```
Comment (leroyvn) [inline eradiate/thermoprops/util.py:287]:
```suggestion
    """
    Interpolates an atmosphere thermophysical properties data set onto a
```

File: eradiate/thermoprops/util.py [hunk 12]
```diff
@@ -32,17 +344,17 @@ def water_vapor_saturation_pressure(t):
     .. note::
         Valid for pressures larger than the triple point pressure (~611 Pa).
 
-    Parameter ``t`` (float):
+    Parameter ``t`` (:class:`float`):
```
Comment (leroyvn) [inline eradiate/thermoprops/util.py:347]: Not required (honestly, nobody looks up the Python reference for the `float` type).

File: eradiate/thermoprops/util.py [hunk 13]
```diff
@@ -83,112 +397,33 @@ def equilibrium_water_vapor_fraction(p, t):
...
-
-    return ds
+            f"{round(p, 2)} Pa, t = {round(t, 2)} K)"
+        )
 
 
 def make_profile_regular(profile, atol):
-    r"""Converts the atmospheric profile with an irregular altitude mesh to a
-    profile defined over a regular altitude mesh
+    """Converts the atmosphere thermophysical properties data set with an
```
Comment (leroyvn) [inline eradiate/thermoprops/util.py:405]:
```suggestion
    """
    Converts the atmosphere thermophysical properties data set with an
```

File: eradiate/thermoprops/util.py [hunk 14]
```diff
@@ -215,32 +450,34 @@ def make_profile_regular(profile, atol):
...
+        attrs=attrs,
     )
     dataset.ert.normalize_metadata(profile_dataset_spec)
 
     return dataset
 
 
 def _to_regular(mesh, atol):
-    r"""Converts an irregular altitude mesh into a regular altitude mesh.
+    """Converts an irregular altitude mesh into a regular altitude mesh.
```
Comment (leroyvn) [inline eradiate/thermoprops/util.py:480]:
```suggestion
    """
    Converts an irregular altitude mesh into a regular altitude mesh.
```

File: eradiate/thermoprops/util.py [hunk 15]
```diff
@@ -252,60 +489,63 @@ def _to_regular(mesh, atol):
...
     """
 
     n, _ = _find_regular_params_gcd(mesh, atol)
     return np.linspace(start=mesh[0], stop=mesh[-1], num=n)
 
 
-def _find_regular_params_gcd(mesh, unit_number=1.):
-    r"""Finds the parameters (number of cells, constant cell width) of the
+def _find_regular_params_gcd(mesh, unit_number=1.0):
+    """Finds the parameters (number of cells, constant cell width) of the
```
Comment (leroyvn) [inline eradiate/thermoprops/util.py:507]:
```suggestion
    """
    Finds the parameters (number of cells, constant cell width) of the
```

File: eradiate/thermoprops/util.py [hunk 16]
```diff
@@ -252,60 +489,63 @@ def _to_regular(mesh, atol):
...
+    Parameter ``unit_number`` (:class:`float`):
         Defines the unit used to convert the floating point numbers to integer.
         numbers.
 
         Default: 1.
 
-    Returns -> int, float:
-        Number of points in the regular mesh and the value of the constant cells
-        width.
+    Returns -> :class:`tuple`:
```
Comment (leroyvn) [inline eradiate/thermoprops/util.py:529]: Please revert to get closer to the tuple typing spec rationale [https://docs.python.org/3/library/typing.html#typing.Tuple].

File: eradiate/data/absorption_spectra.py [hunk 17]
```diff
@@ -48,7 +134,7 @@ def open(cls, id):
             raise OSError(f"while opening file at {path}: {str(e)}")
 
     @classmethod
-    def find(cls):
+    def find(cls, test=False):
```
Comment (leroyvn) [inline eradiate/data/absorption_spectra.py:137]: This the `test` argument used somewhere?

**Review by leroyvn — COMMENTED** [review]
New series of comments. Most are about formatting (easy to add to an automated batch commit).

The absorption coefficient computation code could IMO be made much more concise. I think the mutiple small functions you've created can actually be replaced by in-line code with small comments without losing code legibility.

Regarding docstrings: don't bother wrapping all Python built-in types with `:class:` (saves typing, makes docstring more legible and concise without noticeable loss of functionality).

File: eradiate/data/absorption_spectra.py (hunk 17 above)
Comment (nollety) [reply eradiate/data/absorption_spectra.py:137]: No, that argument is a leftover from a debug session, I forgot to remove it.

File: eradiate/radprops/absorption.py [hunk 18]
```diff
@@ -1,173 +1,214 @@
...
+    else:
+        return xsw.interp(p=pz, t=tz, method=method, kwargs=dict(fill_value=fill_value))
+
+
+def _check_within_range(name, value, lower, upper):
+    value = np.array(value)
+    if (value < lower).any() or (value > upper).any():
+        raise ValueError(
+            f"{name} value(s) ({value}) is outside of range " f"({lower}, {upper})"
+        )
```
[the remark this answers is not in the document]
Comment (nollety) [reply eradiate/radprops/absorption.py:214]: You're right, setting `kwargs=dict(bounds_error=True)` will raise a `ValueError` any time interpolation is attempted on a value outside of the range; I'll refactor the code accordingly!

[the remark this answers is not in the document]
Comment (nollety) [reply eradiate/radprops/absorption.py:214]: The reason I checked manually was that I thought it did not work...
I just checked and the exception is indeed raised.

**Comment by nollety** [comment]
@leroyvn I've applied all of your suggestions and squashed and rebased onto latest main :)

**Comment by leroyvn** [comment]
Thanks, merging now! 🚀

**Pull request merged**